File: projecteuler/go/3/3.go
/*

Largest prime factor

The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?

*/
package main

import (
	"fmt"
	"github.com/zeroshirts/challenges/projecteuler/go/misc"
)

func primeFactors(x uint64) (pf []uint64) {
	for div := range misc.Divisors(x) {
		if misc.IsPrime(div) {
			pf = append(pf, div)
		}
	}
	return pf
}

func largest() uint64 {
	pf := primeFactors(600851475143)
	fmt.Println("prime factors:", pf)
	var x uint64
	for i := range pf {
		if pf[i] > x {
			x = pf[i]
		}
	}
	return x
}

func main() {
	// XXX ??? Why is it that prime factors, when multiplied together, yield target?
	fmt.Println(largest())
}

File: projecteuler/julia/3/3.jl
#=

Largest prime factor

The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?

=#

using Primes
using Test

function largestPrimeFactor(x)
  largest = 0
  for k in keys(factor(x))
    if k > largest
      largest = k
    end
  end
  return largest
end

@time @test factor(13195) == Dict(7=>1, 13 => 1, 29 => 1, 5 => 1)
@time @test largestPrimeFactor(600851475143) == 6857